README.md
<p align="center">
<img src="/assets/timer.png" width="150" height="150" />
</p>

<p align="center">Keep track of your time while working on a particular project</p>

<h1 align="center">Work Stamp</h1>

<p align="center">
    <img src="https://img.shields.io/github/stars/elvis-ndubuisi/work-stamp" alt="GitHub Repo stars"/>
    <img src="https://img.shields.io/github/package-json/v/elvis-ndubuisi/work-stamp" alt="GitHub package.json version (subfolder of monorepo)"/>
    <img src="https://img.shields.io/github/license/elvis-ndubuisi/work-stamp" alt="Github"/>
</p>

Welcome to "Work Stamp," an extension designed to help you keep track of your time while working on specific projects or tasks within Visual Studio Code.

## Commands

- `WK: START/STOP`: starts or stops timer

## Features

- Tracks your active work time and total work time differently.
- Tracks coding activity automatically or via using a command.
- Automatically saves tracked time project when vscode is closed.
- Saves time stamp information locally.
- Display elapsed time in the status bar with tooltip message.

## Extension Settings

- `work-stamp.autoStart`: starts work stamp when typing begins if set to `true`. Default: `true`
- `work-stamp.delayDuration`: delay between typing activities before active work timer pauses(in minutes). Default: `10` minutes.

## TODO

- [ ] Preview all projects timestamp and summary logs in webview.
- [ ] Preview specific project timestamp and summary log
- [ ] Backup log information to custom REST endpoint _(only if specified by a user)_
- [ ] Display active work elapsed time and total elapsed time message tooltip message.
- [ ] Add break reminder

## Requirements

This extension does not have any specific requirements or dependencies.

## Known Issues

There are no known issues at the moment. If you encounter any problems, please [report them on GitHub](https://github.com/elvis-ndubuisi/work-stamp/issues).

## Attributions

[Timer stickers created by Gohsantosadrive - Flaticon](https://www.flaticon.com/free-stickers/timer)

**Enjoy tracking your work time with "Work Stamp"!**


CHANGELOG.md
# Change Log

## 0.2.2

### Patch Changes

- Update extension documentation and package file

## 0.2.1

### Patch Changes

- Added unit tests to utility and extension functions.

## 0.2.0

### Minor Changes

- Added stamp date storage & retrieval
- Preview logs in table using webview
- Bind webview to command : WK: Read
- Track active work duration via typing activities
- Automatically starts timer if auto-start config is true

### Patch Changes

- update extension function for new feature

## 0.1.0

### Minor Changes

- displays extension command in statusBarItem

## 0.0.1 (Release Date: September 4, 2023)

### Patch Changes

- Implemented Work stamp extension MVP.

### Details

- Added the "Work stamp" extension, which serves as the Minimum Viable Product (MVP) for tracking and managing work-related time in Visual Studio Code.
- The extension includes features such as real-time time tracking, start/stop functionality, and the display of elapsed time in the status bar.
- The extension is designed to enhance productivity and time management within the VS Code environment.

Thank you for using Work stamp!


src/test/suite/store.test.ts
import * as store from "../../store";
import * as assert from "assert";
import * as path from "node:path";
import { beforeEach, afterEach } from "mocha";
import * as fs from "node:fs";

suite("Data store utility functions test", () => {
  // Define a temporary directory for testing.
  const tempDir = path.join(__dirname, "temp");
  const tempDatastorePath = path.join(tempDir, "datastore.csv");

  beforeEach(() => {
    // Create a temporary directory before each test.
    fs.mkdirSync(tempDir, { recursive: true });
  });

  afterEach(() => {
    // Cleanup: Delete the temporary directory after each test.
    fs.rmSync(tempDir, { recursive: true });
  });

  test("initStorePath should initialize the datastore directory", () => {
    const datastorePath = store.initStorePath();
    // Verify that the directory is created and the path is correct.
    assert.ok(fs.existsSync(datastorePath!));
  });

  test("writeStampToCsv should create a new CSV file and write stamp data to file", () => {
    // Sample stamp data
    const stamp = {
      date: Date.now(),
      activeDurationStamp: "00:30:00",
      totalDurationStamp: "01:00:00",
      projectName: "Test Project",
      startTime: Date.now(),
      endTime: Date.now() + 3600000, // 1 hour later
    };

    store.writeStampToCsv(tempDatastorePath, stamp);

    // Verify that the file exists and has the correct content.
    assert.ok(fs.existsSync(tempDatastorePath));

    const fileContent = fs.readFileSync(tempDatastorePath, "utf-8");
    const expectedCsvContent =
      "projectName,totalDuration,activeDuration,startTime,endTime,date\n" +
      store.formatCsv(stamp);
    assert.strictEqual(fileContent, expectedCsvContent);
  });

  test("readStampData should read stamp data from CSV file", async () => {
    // Sample stamp data
    const stamp = {
      date: Date.now(),
      activeDurationStamp: "00:30:00",
      totalDurationStamp: "01:00:00",
      projectName: "Test Project",
      startTime: Date.now(),
      endTime: Date.now() + 3600000, // 1 hour later
    };

    const expectedStamp = {
      projectName: stamp.projectName,
      date: new Date(stamp.date).toDateString(),
      startTime: new Date(stamp.startTime).toLocaleTimeString(),
      endTime: new Date(stamp.endTime).toLocaleTimeString(),
      activeDuration: stamp.activeDurationStamp,
      totalDuration: stamp.totalDurationStamp,
    };

    // Create the CSV file with the sample stamp data
    store.writeStampToCsv(tempDatastorePath, stamp);

    // Read the stamp data
    const stampLogs = await store.readStampData(tempDatastorePath);

    // Verify that the data matches the written stamp
    assert.strictEqual(stampLogs.length, 1);
    const readStamp = stampLogs[0];
    assert.deepStrictEqual(readStamp, expectedStamp);
  });

  test("formatCsv should format stamp data into CSV string", () => {
    // Sample stamp data
    const stamp = {
      date: Date.now(),
      activeDurationStamp: "00:30:00",
      totalDurationStamp: "01:00:00",
      projectName: "Test Project",
      startTime: Date.now(),
      endTime: Date.now() + 3600000, // 1 hour later
    };

    const csvString = store.formatCsv(stamp);

    // Verify that the formatted CSV string matches expectations
    const expectedCsvContent = `${stamp.projectName},${
      stamp.totalDurationStamp
    },${stamp.activeDurationStamp},${new Date(
      stamp.startTime
    ).toLocaleTimeString()},${new Date(
      stamp.endTime
    ).toLocaleTimeString()},${new Date(stamp.date).toDateString()}\n`;

    assert.strictEqual(csvString, expectedCsvContent);
  });
});


src/test/suite/utils.test.ts
import * as assert from "assert";
import * as vscode from "vscode";
import {
  getCurrentWorkspaceName,
  formatElapsedTime,
  formatElapsedTimeMsg,
  getElapsedTime,
  joinTimeStamps,
} from "../../utils";

suite("Utility Functions Tests", () => {
  suite("getCurrentWorkspaceName", () => {
    test("should return the current project name", () => {
      // Get the workspace from which the util.test lives in.
      const workspaceName = vscode.window.activeTextEditor;
      const fnWorkspace = getCurrentWorkspaceName();

      assert.strictEqual(workspaceName, fnWorkspace);
    });
  });

  suite("formatElapsedTime", () => {
    test("should format milliseconds into HH:MM:SS", () => {
      const testData = [
        { milliseconds: 0, expected: "00:00:00" },
        { milliseconds: 1000, expected: "00:00:01" },
        { milliseconds: 60000, expected: "00:01:00" },
        { milliseconds: 3600000, expected: "01:00:00" },
        { milliseconds: 3661000, expected: "01:01:01" },
      ];

      testData.forEach((data) => {
        const result = formatElapsedTime(data.milliseconds);
        assert.strictEqual(data.expected, result);
      });
    });
  });

  suite("getElapsedTime", () => {
    test("should return elapsed time in HH:MM:SS format", () => {
      const startTime = Date.now() - 10000;
      const result = getElapsedTime(startTime);

      const matchRegex = /^\d{2}:\d{2}:\d{2}$/;

      assert.match(result, matchRegex);
    });

    test("should return '00:00:00' if startTime is not provided", () => {
      const result = getElapsedTime(23256754.32456);
      assert.notEqual(result, "00:00:00");
    });
  });

  suite("formatElapsedTimeMsg", () => {
    test("should return 'You've worked for less than on minute' and typeOf string", () => {
      const alert = formatElapsedTimeMsg("00:00:00");
      assert.deepStrictEqual(typeof alert, typeof "");
      assert.strictEqual("You've worked for less than a minute", alert);
    });

    test("should return 'You've worked for 1 minute'", () => {
      const alert = formatElapsedTimeMsg("00:01:023");
      assert.strictEqual("You've worked for 1 minute", alert);
    });

    test("should return a plural minute 'You've worked for 1 minutes'", () => {
      const alert = formatElapsedTimeMsg("00:20:023");
      assert.strictEqual("You've worked for 20 minutes", alert);
    });

    test("should not return 'You've worked for 20 hour'", () => {
      const alert = formatElapsedTimeMsg("20:01:00");
      assert.notStrictEqual("You've worked for 20 hours", alert);
    });

    test("should return not return 'You've worked for 20 hour and 20 minutes'", () => {
      const alert = formatElapsedTimeMsg("20:20:023");
      assert.strictEqual("You've worked for 20 hours and 20 minutes", alert);
    });
  });

  suite("joinTimeStamps", () => {
    test("should correctly join two time stamps", () => {
      const testData = [
        { stamp1: "00:00:00", stamp2: "00:00:01", expected: "00:00:01" },
        { stamp1: "00:00:59", stamp2: "00:00:01", expected: "00:01:00" },
        { stamp1: "00:59:59", stamp2: "00:00:01", expected: "01:00:00" },
        { stamp1: "01:59:59", stamp2: "00:00:01", expected: "02:00:00" },
        { stamp1: "23:59:59", stamp2: "00:00:01", expected: "24:00:00" },
      ];

      testData.forEach((data) => {
        const result = joinTimeStamps(data.stamp1, data.stamp2);
        assert.strictEqual(result, data.expected);
      });
    });
  });
});


src/test/suite/extension.test.ts
import * as assert from "assert";
import * as vscode from "vscode";
import * as extension from "../../extension";
import { before } from "mocha";

// TODO: revisit test suites

suite("Extension Test Suite", () => {
  vscode.window.showInformationMessage("Start all tests.");

  let context: vscode.ExtensionContext;

  before(() => {
    context = {
      extensionPath: "/mock-path",
      subscriptions: [],
    } as unknown as vscode.ExtensionContext; // Create a mock content object.
  });

  // test("activate function should initialize the extension", async () => {
  //   await extension.activate(context); // Call your activate function

  //   // Add assertions here to check if your extension has been initialized properly.
  //   // For example, you can assert that the status bar item has been created.
  //   assert.ok(extension.statusBarItem);
  // });

  test("deactivate function should clean up resources", () => {
    extension.deactivate(); // Call your deactivate function

    // Add assertions here to check if your extension's resources have been cleaned up properly.
  });

  test("startStampTimer should start the timer", () => {
    extension.startStampTimer(); // Call your startStampTimer function

    // Add assertions here to check if the timer has been started correctly.
    assert.strictEqual(extension.isTimerRunning, true);
  });

  test("stopStampTimer should stop the timer", () => {
    extension.stopStampTimer(); // Call your stopStampTimer function

    // Add assertions here to check if the timer has been stopped correctly.
    assert.strictEqual(extension.isTimerRunning, false);
  });

  // test("workStamp command should start or stop the timer", async () => {
  //   // Ensure the timer is not running initially
  //   // extension.isTimerRunning = false;

  //   // Call the workStamp command
  //   await vscode.commands.executeCommand(extension.cmdIds.start);

  //   // Add assertions here to check if the timer has been started.

  //   // Call the workStamp command again
  //   await vscode.commands.executeCommand(extension.cmdIds.start);

  //   // Add assertions here to check if the timer has been stopped.
  // });

  teardown(() => {
    vscode.window.showInformationMessage("All tests have finished.");
  });
});


src/store.ts
import * as fs from "node:fs";
import * as path from "node:path";
import * as vscode from "vscode";
import * as csv from "csv-parser";

type IStamp = {
  date: number;
  activeDurationStamp: string;
  totalDurationStamp: string;
  projectName: string;
  startTime: number;
  endTime: number;
};

const EXT_ID = "ElvisNdubuisi.work-stamp";

export function initStorePath(): string | null {
  const extensionPath = vscode.extensions.getExtension(EXT_ID)!.extensionPath;
  let dataFolderPath = path.join(extensionPath, ".stamp-data");

  // Make dir if none exists
  if (!fs.existsSync(dataFolderPath)) {
    fs.mkdirSync(dataFolderPath);
    vscode.window.showInformationMessage("Initialized datastore directory");
  }

  // Define file path.
  const filePath = path.join(dataFolderPath, "datastore.csv");

  // Create the CSV file if it doesn't exist
  if (!fs.existsSync(filePath)) {
    fs.writeFileSync(
      filePath,
      "projectName,totalDuration,activeDuration,startTime,endTime,date\n",
      "utf-8"
    );
  }

  return filePath;
}

export function writeStampToCsv(datastore: string, stamp: IStamp): void {
  // Check if datastore has been created.
  if (!fs.existsSync(datastore)) {
    fs.writeFileSync(
      datastore,
      "projectName,totalDuration,activeDuration,startTime,endTime,date\n",
      "utf-8"
    );
  }

  // Append stamp data to data-source
  fs.appendFileSync(datastore, formatCsv(stamp));
}

export function readStampData(datastore: string): Promise<any[]> {
  return new Promise((resolve, reject) => {
    const stampLogs: IStamp[] = [];

    fs.createReadStream(datastore)
      .pipe(csv())
      .on("data", (log) => {
        stampLogs.push(log);
      })
      .on("end", () => {
        resolve(stampLogs);
      })
      .on("error", (err: any) => {
        reject(err);
      });
  });
}

export function formatCsv(stamp: IStamp): string {
  const date = new Date(stamp.date).toDateString();
  const startTime = new Date(stamp.startTime).toLocaleTimeString();
  const endTime = new Date(stamp.endTime).toLocaleTimeString();
  return `${stamp.projectName},${stamp.totalDurationStamp},${stamp.activeDurationStamp},${startTime},${endTime},${date}\n`;
}


src/utils.ts
import * as vscode from "vscode";

// Get the name of the active workspace
export function getCurrentWorkspaceName(): string | undefined {
  const activeTextEditor = vscode.window.activeTextEditor; // Get the current active text editor
  const workspaceFolders = vscode.workspace.workspaceFolders; // Get the list of workspace folders in the currently open workspace
  let activeWorkspaceName: string | undefined = undefined;

  if (activeTextEditor) {
    // Get the workspace folder associated with the active text editor if any
    const activeWorkspace = vscode.workspace.getWorkspaceFolder(
      activeTextEditor.document.uri
    );
    if (activeWorkspace) {
      activeWorkspaceName = activeWorkspace.name; // Return the name of workspace if any was found
    }
  } else if (workspaceFolders && workspaceFolders.length > 0) {
    // If there are workspace folders, return the name of the first one.
    activeWorkspaceName = workspaceFolders[0].name;
  }
  //   Returns undefined if no active text editor or associated workspace folder is found
  return activeWorkspaceName;
}

export function formatElapsedTime(milliseconds: number): string {
  const seconds = Math.floor((milliseconds / 1000) % 60);
  const minutes = Math.floor((milliseconds / 1000 / 60) % 60);
  const hours = Math.floor(milliseconds / 1000 / 60 / 60);
  return `${padZero(hours)}:${padZero(minutes)}:${padZero(seconds)}`;
}

export function getElapsedTime(startTime: number): string {
  if (startTime) {
    const currentTime = Date.now();
    const elapsedTime = formatElapsedTime(currentTime - startTime);
    return elapsedTime;
  }
  return "00:00:00";
}

export function formatElapsedTimeMsg(timestamp: string): string {
  // Split the timestamp into hours, minutes, and seconds
  const [hoursStr, minutesStr, secondsStr] = timestamp.split(":");

  // Convert the parts to numbers
  const hours = parseInt(hoursStr, 10);
  const minutes = parseInt(minutesStr, 10);
  const seconds = parseInt(secondsStr, 10);

  // Calculate the total number of seconds
  const totalSeconds = hours * 3600 + minutes * 60 + seconds;

  // Calculate hours, minutes, and remaining seconds
  const formattedHours = Math.floor(totalSeconds / 3600);
  const formattedMinutes = Math.floor((totalSeconds % 3600) / 60);

  // Create the formatted text
  let formattedText = "You've worked for ";

  if (formattedHours > 0) {
    formattedText += `${formattedHours} hour${formattedHours > 1 ? "s" : ""}`;
  }

  if (formattedMinutes > 0) {
    if (formattedHours > 0) {
      formattedText += " and ";
    }
    formattedText += `${formattedMinutes} minute${
      formattedMinutes > 1 ? "s" : ""
    }`;
  }

  if (formattedHours === 0 && formattedMinutes === 0) {
    formattedText += "less than a minute";
  }

  return formattedText;
}

export function joinTimeStamps(stamp1: string, stamp2: string): string {
  // Split the timestamps into hours, minutes, and seconds
  const [hours1, minutes1, seconds1] = stamp1.split(":").map(Number);
  const [hours2, minutes2, seconds2] = stamp2.split(":").map(Number);

  // Calculate the total seconds
  const totalSeconds =
    seconds1 +
    seconds2 +
    minutes1 * 60 +
    minutes2 * 60 +
    hours1 * 3600 +
    hours2 * 3600;

  // Calculate the new hours, minutes, and seconds
  const newHours = Math.floor(totalSeconds / 3600);
  const remainingSeconds = totalSeconds % 3600;
  const newMinutes = Math.floor(remainingSeconds / 60);
  const newSeconds = remainingSeconds % 60;

  // Format the result as "HH:MM:SS"
  const result = `${String(newHours).padStart(2, "0")}:${String(
    newMinutes
  ).padStart(2, "0")}:${String(newSeconds).padStart(2, "0")}`;

  return result;
}

function padZero(num: number): string {
  return num.toString().padStart(2, "0");
}


src/views.ts
import * as vscode from "vscode";

type LogsType = {
  date: string;
  projectName: string;
  totalDuration: string;
  activeDuration: string;
  startTime: string;
  endTime: string;
};

export function getHtmlForWebView(
  webview: vscode.Webview,
  context: vscode.ExtensionContext,
  logs: any,
  projectName?: string
) {
  // Get local paths to assets, convert to special URI for webview.
  const mainScriptPath = webview.asWebviewUri(
    vscode.Uri.joinPath(context.extensionUri, "assets", "main.js")
  );
  const vsStylesPath = webview.asWebviewUri(
    vscode.Uri.joinPath(context.extensionUri, "assets", "vscode.css")
  );
  const mainStylesPath = webview.asWebviewUri(
    vscode.Uri.joinPath(context.extensionUri, "assets", "main.css")
  );

  return `<!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="UTF-8" />
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <title>Work Stamp Logs</title>
      <link rel="stylesheet" href="${mainStylesPath}" />
      <link rel="stylesheet" href="${vsStylesPath}" />
      <meta
        http-equiv="Content-Security-Policy"
        content="default-src 'none'; img-src ${
          webview.cspSource
        } https script-src ${webview.cspSource}; frame-src 'self'; style-src ${
    webview.cspSource
  };"
      />
    </head>
    <body>
      <h1 class="heading">Heading</h1>
      <section class="table_container">
        <table>
          <thead>
            <td>Date</td>
            <td>Project Name</td>
            <td>Duration</td>
            <td>Active</td>
            <td>Start Time</td>
            <td>End Time</td>
          </thead>
          <tbody>${spreadLogs(logs, projectName && projectName)}</tbody>
        </table>
      </section>
      <script src="${mainScriptPath}"></script>
    </body>
  </html>`;
}

function spreadLogs(logs: LogsType[], filter?: string): string | null {
  const html = logs.map((log) => {
    if (filter) {
      if (filter === log.projectName) {
        return `<tr>
          <td>${log.date}</td>
          <td>${log.projectName}</td>
          <td>${log.totalDuration}</td>
          <td>${log.activeDuration}</td>
          <td>${log.startTime}</td>
          <td>${log.endTime}</td>
        </tr>`;
      } else {
        throw Error("Project not found");
      }
    }
    return `<tr>
      <td>${log.date}</td>
      <td>${log.projectName}</td>
      <td>${log.totalDuration}</td>
      <td>${log.activeDuration}</td>
      <td>${log.startTime}</td>
      <td>${log.endTime}</td>
    </tr>`;
  });
  return `${html}`;
}


src/extension.ts
import * as vscode from "vscode";
import {
  getCurrentWorkspaceName,
  formatElapsedTime,
  getElapsedTime,
  formatElapsedTimeMsg,
  joinTimeStamps,
} from "./utils";
import { initStorePath, writeStampToCsv, readStampData } from "./store";
import { genWebViewContent } from "./views";

export let datastore: string | null = "";
let projectName: string | undefined = "";
let startTime: number | null = null;
let activeStartTime: number | null = null;
let activeDuration: string = "00:00:00";
export let statusBarItem: vscode.StatusBarItem | null = null;
let intervalId: NodeJS.Timer | null = null;
let delayIntervalId: NodeJS.Timer | null = null;
let delayTimerId: NodeJS.Timeout | null = null;
export let isTimerRunning: boolean = false;
let hasStartedCoding: boolean = false;
let delayDuration: number | null = 0;

const currentWorkspace = getCurrentWorkspaceName();
export const cmdIds = {
  start: "work-stamp.stamp-work", // start stamp timer
  read: "work-stamp.stamp-read", // read stamp data
  project: "work-stamp.stamp-project", // read project stamp data
};
const settingIds = {
  auto: "work-stamp.autoStart", // timer auto start boolean from settings
  delay: "work-stamp.delayDuration", // delay duration from settings
};

// Your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  console.log('Congratulations, your extension "work-stamp" is now active!');

  datastore = initStorePath(); // Initialize stamp data storage source.
  let iconPath = vscode.Uri.joinPath(
    context.extensionUri,
    "assets",
    "timer.png"
  );

  // Read workspace configurations.
  const config = vscode.workspace.getConfiguration();
  const canAutoStart = config.get(settingIds.auto, true);
  const delayMinutes = config.get(settingIds.delay, 4);
  delayDuration = delayMinutes * 60 * 1000; // convert delay duration to milliseconds.

  // Create status bar item.
  statusBarItem = vscode.window.createStatusBarItem(
    vscode.StatusBarAlignment.Right,
    80
  );
  statusBarItem.command = cmdIds.start;

  // Show statusBarItem if a workspace is active
  if (!currentWorkspace) {
    vscode.window.showErrorMessage(
      "No project or workspace is open. Start the timer with an open project"
    );
    return;
  } else {
    projectName = currentWorkspace;
    statusBarItem.show();
  }

  // Register command to start/stop timer - via statusBarItem or Command.
  const workStamp = vscode.commands.registerCommand(cmdIds.start, () => {
    if (isTimerRunning) {
      // Save data and stop timer
      if (projectName && startTime) {
        const endTime = Date.now();
        const elapsedTime = formatElapsedTime(endTime - startTime);
        if (datastore) {
          writeStampToCsv(datastore, {
            activeDurationStamp: activeDuration ? activeDuration : "00:00:00",
            date: Date.now(),
            endTime: endTime,
            projectName: projectName,
            startTime: startTime,
            totalDurationStamp: elapsedTime,
          });
          vscode.window.showInformationMessage("🚀Saved timestamp🕥.");
        }
      }
      stopStampTimer();
    } else {
      // Start timer
      startTime = Date.now();
      startStampTimer();
    }
    updateStatusBarItem();
    updateActiveWorkTime();
  });

  // Listen for active work-time via document change
  const checkActiveWorkTime = vscode.workspace.onDidChangeTextDocument(() => {
    // Check if work stamp is not running & auto-start is enabled.
    if (!isTimerRunning && canAutoStart) {
      startTime = Date.now();
      startStampTimer();
    }

    if (!hasStartedCoding) {
      // Start the timer when coding activity begins - without auto start
      hasStartedCoding = true;
      activeStartTime = Date.now();
    }
  });

  // Register command to view timestamp logs.
  const readStamp = vscode.commands.registerCommand(cmdIds.read, () => {
    const panel = vscode.window.createWebviewPanel(
      "workStamp",
      "Work Stamp",
      vscode.ViewColumn.One,
      {
        enableScripts: true,
        localResourceRoots: [
          vscode.Uri.joinPath(context.extensionUri, "views"),
          vscode.Uri.joinPath(context.extensionUri, "assets"),
        ],
      }
    );
    panel.iconPath = iconPath;

    // Read Log stamps.
    if (!datastore) {
      vscode.window.showErrorMessage("No data source found");
      return;
    }

    readStampData(datastore)
      .then((logs) => {
        panel.webview.html = genWebViewContent(logs, {
          icon: panel.webview.asWebviewUri(iconPath),
          style: panel.webview.asWebviewUri(stylePath),
          app: panel.webview.asWebviewUri(scriptPath),
          cspSource: panel.webview.cspSource,
        });
      })
      .catch((err) => {
        vscode.window.showErrorMessage(err);
      });
  });

  const readProject = vscode.commands.registerCommand(cmdIds.project, () => {
    const panel = vscode.window.createWebviewPanel(
      "",
      `projectName`,
      vscode.ViewColumn.One
    );
  });

  // Update status bar item initially
  updateStatusBarItem();

  // Subscription contexts.
  context.subscriptions.push(statusBarItem);
  context.subscriptions.push(
    workStamp,
    readStamp,
    readProject,
    checkActiveWorkTime
  );
}

// This method is called when your extension is deactivated
export function deactivate() {
  // Save Stamp - (unexpected deactivation)
  if (isTimerRunning && startTime) {
    const endTime = Date.now();
    const elapsedTime = formatElapsedTime(endTime - startTime);
    if (datastore) {
      writeStampToCsv(datastore, {
        activeDurationStamp: activeDuration ? activeDuration : "00:00:00",
        date: Date.now(),
        endTime: endTime,
        projectName: projectName!,
        startTime: startTime,
        totalDurationStamp: elapsedTime,
      });
    }
  }

  stopStampTimer();
}

function updateStatusBarItem(): void {
  // Show formatted message tooltip
  statusBarItem!.tooltip = formatElapsedTimeMsg(
    startTime ? getElapsedTime(startTime) : "00:00:00"
  );

  if (isTimerRunning) {
    statusBarItem!.text = `$(flame) ${
      startTime ? getElapsedTime(startTime) : "00:00:00"
    }`;
  } else {
    statusBarItem!.text = `$(watch) Start Stamp`;
  }
  statusBarItem!.show();
}

export function startStampTimer(): void {
  if (!intervalId) {
    isTimerRunning = true;
    intervalId = setInterval(updateStatusBarItem, 1000);
  }

  if (!delayIntervalId) {
    hasStartedCoding = false;
    delayIntervalId = setInterval(updateActiveWorkTime, 1000);
  }
}

export function stopStampTimer(): void {
  if (intervalId) {
    clearInterval(intervalId);
    intervalId = null;
    isTimerRunning = false;
    startTime = null;
  }

  if (delayTimerId) {
    clearTimeout(delayTimerId);
    delayTimerId = null;
  }
  if (delayIntervalId) {
    clearInterval(delayIntervalId);
    delayIntervalId = null;
  }

  hasStartedCoding = false;
  activeStartTime = null;
  activeDuration = "00:00:00";
}

function updateActiveWorkTime(): void {
  if (hasStartedCoding && activeStartTime) {
    const currentTime = Date.now();
    const elapsedTime = formatElapsedTime(currentTime - activeStartTime);

    // Update activeDuration - added elapsed time to previous activeDuration
    activeDuration = joinTimeStamps(activeDuration, elapsedTime);

    // Reset the activeStartTime.
    activeStartTime = currentTime;

    // Prevent further activeDuration if time of last keystroke > delayDuration.
    delayTimerId = setTimeout(() => {
      hasStartedCoding = false;
    }, delayDuration!);
  }
}
